sale_stock_propagate_quantity_diminution/tests/test_sale_stock_propagate_quantity_diminution.py

```diff
@@ -0,0 +1,255 @@
+        wiz = Form(
+            self.env[wiz_act["res_model"]].with_context(wiz_act["context"])
+        ).save()
+        wiz.process()
+        pick2 = self.order4.picking_ids.filtered(
+            lambda r: r.location_id.id == 8 and r.state not in ["cancel", "done"]
+        )
+        self.assertEqual(len(self.order4.picking_ids), 4)
+        self.assertTrue(pick1.state == "done")
+        self.assertTrue(pick2.state == "confirmed")
+        stock_move_pick2 = stock_move_pack1.move_orig_ids.filtered(
+            lambda r: r.state not in ["cancel", "done"]
+        )
+        self.assertEqual(stock_move_pick2.product_uom_qty, 3)
+        # partial delivery of the pack
+        pack1 = self.order4.picking_ids.filtered(
+            lambda r: r.location_id.id == 12 and r.state not in ["cancel", "done"]
+        )
+        stock_move_pack1.write({"quantity_done": 8})
+        wiz_act = pack1.button_validate()
+        wiz = Form(
+            self.env[wiz_act["res_model"]].with_context(wiz_act["context"])
+        ).save()
+        wiz.process()
+        pack2 = self.order4.picking_ids.filtered(
+            lambda r: r.location_id.id == 12 and r.state not in ["cancel", "done"]
+        )
+        self.assertEqual(len(self.order4.picking_ids), 5)
+        self.assertTrue(pack1.state == "done")
+        self.assertTrue(pack2.state == "waiting")
+        stock_move_pack2 = stock_move_out1.move_orig_ids.filtered(
+            lambda r: r.state not in ["cancel", "done"]
+        )
+        self.assertEqual(stock_move_pack2.product_uom_qty, 5)
+        # partial delivery of the out
+        out1 = self.order4.picking_ids.filtered(
+            lambda r: r.location_id.id == 11 and r.state not in ["cancel", "done"]
+        )
+        stock_move_out1.write({"quantity_done": 6})
+        wiz_act = out1.button_validate()
+        wiz = Form(
+            self.env[wiz_act["res_model"]].with_context(wiz_act["context"])
+        ).save()
+        wiz.process()
+        out2 = self.order4.picking_ids.filtered(
+            lambda r: r.location_id.id == 11 and r.state not in ["cancel", "done"]
+        )
+        self.assertEqual(len(self.order4.picking_ids), 6)
+        self.assertTrue(out1.state == "done")
+        self.assertTrue(out2.state == "waiting")
+        stock_move_out2 = line.move_ids.filtered(
+            lambda r: r.state not in ["cancel", "done"]
+        )
+        self.assertEqual(stock_move_out2.product_uom_qty, 7)
+        # decrease of quantities
+        line.write({"product_uom_qty": 10})
+        self.assertEqual(line.product_uom_qty, 10)
+        self.assertEqual(stock_move_pick1.product_uom_qty, 10)
+        self.assertEqual(stock_move_pack1.product_uom_qty, 8)
+        self.assertEqual(stock_move_out1.product_uom_qty, 6)
+        self.assertEqual(stock_move_pick2.product_uom_qty, 0)
+        self.assertEqual(stock_move_pack2.product_uom_qty, 2)
+        self.assertEqual(stock_move_out2.product_uom_qty, 4)
```

Comment:
Your test are really really verbose, too much thing (code repetition, useless assert).
Normally a test or a python function should have 20 lines of code, (you can have exception), but the idea is that when is really really long is just unreadable, and we write the code one time and we read it 1000 times, so please make them shorted

Reply:
@sebastienbeau , I have simplified the tests.